contracts/src/tokens/memecoin.cairo

```diff
@@ -91,12 +108,75 @@ mod UnruggableMemecoin {
         // ************************************
         // * UnruggableMemecoin functions
         // ************************************
-        fn launch_memecoin(ref self: ContractState) {
-            // Checks: Only the owner can launch the memecoin.
+
+        /// Launches Memecoin by creating a liquidity pool with the specified counterparty token using the AMMv2 protocol.
+        /// The owner must send both MT tokens (Memecoin) and tokens of the chosen counterparty (e.g., USDC) to launch memecoin.
+        ///
+        /// # Arguments
+        /// - `amm_v2`: AMMV2 to create pair and send liquidity
+        /// - `counterparty_token_address`: The contract address of the counterparty token.
+        /// - `liquidity_memecoin_amount`: The amount of Memecoin tokens to be provided as liquidity.
+        /// - `liquidity_counterparty_token`: The amount of counterparty tokens to be provided as liquidity.
+        ///
+        /// # Panics
+        /// This method will panic if:
+        /// - The caller is not the owner of the contract.
+        /// - Insufficient Memecoin funds are available for liquidity.
+        /// - Insufficient counterparty token funds are available for liquidity.
+        fn launch_memecoin(
+            ref self: ContractState,
+            amm_v2: AMMV2,
+            counterparty_token_address: ContractAddress,
+            liquidity_memecoin_amount: u256,
+            liquidity_counterparty_token: u256,
+        ) {
+            // [Check Owner] Only the owner can launch the Memecoin
             self.ownable.assert_only_owner();
-        // Effects.
 
-        // Interactions.
+            let memecoin_address = starknet::get_contract_address();
+            let caller_address = get_caller_address();
+
+            // [Create Pool]
+            let amm_router = IRouterC1Dispatcher {
+                contract_address: self.amm_configs.read(amm_v2.into()),
+            };
+            assert(amm_router.contract_address.is_non_zero(), 'AMM not supported');
+
+            let amm_factory = IFactoryC1Dispatcher { contract_address: amm_router.factory(), };
+            let pair_address = amm_factory
+                .create_pair(counterparty_token_address, memecoin_address);
+
+            // [Check Balance]
+            let memecoin_balance = self.balance_of(memecoin_address);
+            let counterparty_token_dispatcher = IERC20Dispatcher {
+                contract_address: counterparty_token_address,
+            };
+            let counterparty_token_balance = counterparty_token_dispatcher
+                .balance_of(memecoin_address);
```

Comment:
`balance_of` will not be supported on most coins (e.g. ETH) - use camelCase selector here